src/handlers/users.rs

```
@@ -0,0 +1,73 @@
+use warp::Filter;
+use std::sync::Arc;
+use tokio::sync::RwLock;
+use warp::reply::Reply;
+use std::str::FromStr;
+use crate::users::user::{UserDB, UserRegistration, UserCredentials, SessionToken, Session};
+use crate::error::Result;
+
+type DB<T> = Arc<RwLock<T>>;
+
+pub fn authenticate<T: UserDB>(user_db: DB<T>) -> impl warp::Filter<Extract = (Session,), Error = warp::Rejection> + Clone {
+    async fn do_authenticate<T: UserDB>(user_db: DB<T>, token: String) -> Result<Session, warp::Rejection>  {
+        let token = SessionToken::from_str(&token).map_err(|_| warp::reject())?;
+        let db = user_db.read().await;
+        db.session(token).map_err(|_| warp::reject())
+    }
+
+    warp::any()
+        .and(warp::any().map(move || Arc::clone(&user_db)))
+        .and(warp::header::<String>("authorization"))
+        .and_then(do_authenticate)
+}
+
+pub fn register_user_handler<T: UserDB>(user_db: DB<T>) -> impl Filter<Extract = impl warp::Reply, Error = warp::Rejection> + Clone  {
+    warp::post()
+        .and(warp::path!("user" / "register"))
+        .and(warp::body::json())
+        .and(warp::any().map(move || Arc::clone(&user_db)))
+        .and_then(register_user)
+}
+
+// TODO: move into handler once async closures are available?
+async fn register_user<T: UserDB>(user: UserRegistration, user_db: DB<T>) -> Result<impl warp::Reply, warp::Rejection> {
+    let mut db = user_db.write().await;
+    match db.register(user) {
+        Ok(id) => Ok(warp::reply::json(&id).into_response()),
+        Err(_) => Ok(warp::http::StatusCode::BAD_REQUEST.into_response())
```

You basically ignore the error, although it could be used for providing error information to the user.

---

89ce33d

---

Not quite, my suggestion:
```
let id = db.register(user)?;
Ok(warp::reply::json(&id).into_response())
```

---

cab5afc